### PR TITLE
QE: Adding support for AOCC 4.0 and zen4

### PR DESCRIPTION
**Pull request by amd-toolchain-support**
This pull request adds support for AOCC compiler for QE 7.1 and fixes compatibility between AOCC and QE 7.0

### CODE REVIEW

File: var/spack/repos/builtin/packages/quantum-espresso/package.py [hunk 1]
```diff
@@ -367,7 +367,8 @@ class QuantumEspresso(CMakePackage, Package):
     )
 
     # Configure updated to work with AOCC compilers
-    patch("configure_aocc.patch", when="@6.7:6.8 %aocc")
+    patch("configure_aocc.patch", when="@6.7:7.0 %aocc")
+    patch("configure_qe-7.1_aocc.patch", when="@7.1: %aocc")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/quantum-espresso/package.py:371]: Did you upstream this patch?

File: var/spack/repos/builtin/packages/quantum-espresso/package.py [hunk 2]
```diff
@@ -590,11 +591,19 @@ def install(self, pkg, spec, prefix):
                 zlib_libs = spec["zlib"].prefix.lib + " -lz"
                 filter_file(zlib_libs, format(spec["zlib"].libs.ld_flags), make_inc)
 
+        # Filer file must be applied for AOCC due to incorrect make.inc generation
+        if spec.satisfies("%aocc") and spec.satisfies("@6.8:"):
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/quantum-espresso/package.py]:
```suggestion
        if spec.satisfies("@6.8: %aocc"):
```

**Review by alalazo — COMMENTED** [review]
Just a couple of minor questions. Waiting for maintainers review before merging.

File: var/spack/repos/builtin/packages/quantum-espresso/package.py (hunk 1 above)
Comment (amd-toolchain-support) [reply var/spack/repos/builtin/packages/quantum-espresso/package.py:371]: No, this patch has not been upstreamed to QE codebase yet.

Comment (alalazo) [reply var/spack/repos/builtin/packages/quantum-espresso/package.py:371]: Can you please submit it, so we don't have to rely on patches on the next version?

File: var/spack/repos/builtin/packages/quantum-espresso/configure_qe-7.1_aocc.patch [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+ ppc64:*xlf* | ppc64le:*xlf* )
+         try_cc="xlc_r $try_cc"
+         ;;
+@@ -3763,11 +3773,14 @@ ppc64-bg:* )
+ ppc64-bgq:* )
+         try_cflags="-O3"
+         ;;
++x86_64:* )
++        try_cflags="-Ofast -Mstack_arrays"
++        try_dflags="-D__OPENMP"
```
Comment (ye-luo) [inline var/spack/repos/builtin/packages/quantum-espresso/configure_qe-7.1_aocc.patch:42]: OpenMP is optional. It should not be hard coded.

File: var/spack/repos/builtin/packages/quantum-espresso/package.py [hunk 4]
```diff
@@ -590,11 +591,19 @@ def install(self, pkg, spec, prefix):
...
+            filter_file("FOX_FLAGS = -D__PGI", "FOX_FLAGS = -D__PGI -cpp", make_inc)
+
         # QE 6.8 and later has parallel builds fixed
         if spec.satisfies("@:6.7"):
             parallel_build_on = False
         else:
             parallel_build_on = True
+        # Parallel build fails with 7.1
+        if spec.satisfies("@7.1"):
+            parallel_build_on = False
```
Comment (ye-luo) [inline var/spack/repos/builtin/packages/quantum-espresso/package.py:606]: Why this needs to be special. Are you saying parallel build is broken with any compiler?

**Review by ye-luo — CHANGES_REQUESTED** [review]
The necessary patch needs to proof itself by being accepted by upstream q-e at gitlab. Then we can keep a backport here for 7.1.

Q: does the default CMake work with AOCC 4.0? Older AOCC worked when I manually build QE via CMake. Wondering if AOCC 4.0 changes anything.

**Comment by prckent** [comment]
As a QE and spack user I am happy to see this work on getting optimized AMD support - QE has a large userbase. Echoing ye-luo's question, since the preferred QE configuration/installation route is clearly now CMake, are any updates needed there or does the CMake route already work optimally with AMD compilers and libraries?

**Comment by ye-luo** [comment]
I tried a manual build out side spack with cmake. I don't have MPI enabled.
```
cmake -DCMAKE_C_COMPILER=/soft/AMD/aocc-compiler-4.0.0/bin/clang -DCMAKE_Fortran_COMPILER=/soft/AMD/aocc-compiler-4.0.0/bin/flang -DCMAKE_PREFIX_PATH=/soft/AMD/aocl-linux-aocc/4.0/lib -DQE_ENABLE_MPI=OFF -DQE_ENABLE_OPENMP=ON -DFFTW3_ROOT=/soft/AMD/aocl-linux-aocc/4.0 ../q-e-repo
```
I noticed that OpenMP is required to be on because libflame provided by AOCL requires it.
So with cmake there is no patching needed on QE source code.

**Comment by haampie** [comment]
This is yet another case where AMD's magical compiler performance comes from ... `-Ofast`.

`-Ofast` implies `-ffast-math` (at least on clang, on which aocc is based), and `-ffast-math` enables flush to zero, which can influence correctness of other, unrelated programs, because it's a toggle on the CPU (at least on Intel CPUs): https://simonbyrne.github.io/notes/fastmath/#flushing_subnormals_to_zero

Are AMD folks aware of this? Is this intended?

See https://simonbyrne.github.io/notes/fastmath/#flushing_subnormals_to_zero for example.

**Comment by amd-toolchain-support** [comment]
This patch was internal and should not have made it into the queue for publication.  Our apologies for the noise.  A cleanup of our existing patching will be incoming as a new MR.

**Pull request closed without merging**